#! /bin/sh /usr/share/dpatch/dpatch-run
## opt-44_rotor.dpatch by Thomas Bergwinkl <Bergwinkl.Thomas@vr-web.de>
## extracted from the rotor plugin 0.1.4-vdr1.5
## original filename: vdr-1.5.5-rotor.diff
##
## Thomas Günther <tom@toms-cafe.de>:
##   - made compatible to VDR-1.4.7
##   - adapted to VDR-1.7.8
##   - adapted to VDR-1.7.11
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: This patch is needed for the rotor plugin.

@DPATCH@
diff -Naurp vdr-1.7.11/device.h vdr-1.7.11-rotor/device.h
--- vdr-1.7.11/device.h	2010-01-01 15:04:27.000000000 +0000
+++ vdr-1.7.11-rotor/device.h	2010-01-07 23:49:39.000000000 +0000
@@ -24,6 +24,8 @@
 #include "spu.h"
 #include "thread.h"
 #include "tools.h"
+#include <asm/types.h>
+#include <linux/dvb/frontend.h>
 
 #define MAXDEVICES         16 // the maximum number of devices in the system
 #define MAXPIDHANDLES      64 // the maximum number of different PIDs per device
@@ -269,6 +271,7 @@ public:
   virtual bool HasProgramme(void);
          ///< Returns true if the device is currently showing any programme to
          ///< the user, either through replaying or live.
+  virtual bool SendDiseqcCmd(dvb_diseqc_master_cmd cmd) {return false;}
 
 // PID handle facilities
 
diff -Naurp vdr-1.7.11/dvbdevice.c vdr-1.7.11-rotor/dvbdevice.c
--- vdr-1.7.11/dvbdevice.c	2010-01-04 14:06:24.000000000 +0000
+++ vdr-1.7.11-rotor/dvbdevice.c	2010-01-07 23:55:35.000000000 +0000
@@ -30,6 +30,7 @@
 class cDvbTuner : public cThread {
 private:
   enum eTunerStatus { tsIdle, tsSet, tsTuned, tsLocked };
+  bool SendDiseqc;
   int fd_frontend;
   int adapter, frontend;
   int tuneTimeout;
@@ -42,6 +43,7 @@ private:
   cMutex mutex;
   cCondVar locked;
   cCondVar newSet;
+  dvb_diseqc_master_cmd diseqc_cmd;
   bool GetFrontendStatus(fe_status_t &Status, int TimeoutMs = 0);
   bool SetFrontend(void);
   virtual void Action(void);
@@ -50,11 +52,13 @@ public:
   virtual ~cDvbTuner();
   bool IsTunedTo(const cChannel *Channel) const;
   void Set(const cChannel *Channel);
+  bool SendDiseqcCmd(dvb_diseqc_master_cmd cmd);
   bool Locked(int TimeoutMs = 0);
   };
 
 cDvbTuner::cDvbTuner(int Fd_Frontend, int Adapter, int Frontend, fe_delivery_system FrontendType)
 {
+  SendDiseqc=false;
   fd_frontend = Fd_Frontend;
   adapter = Adapter;
   frontend = Frontend;
@@ -123,6 +127,17 @@ bool cDvbTuner::Locked(int TimeoutMs)
   return tunerStatus >= tsLocked;
 }
 
+bool cDvbTuner::SendDiseqcCmd(dvb_diseqc_master_cmd cmd)
+{
+  cMutexLock MutexLock(&mutex);
+  if ((frontendType != SYS_DVBS && frontendType != SYS_DVBS2) || SendDiseqc)
+    return false;
+  diseqc_cmd=cmd;
+  SendDiseqc=true;
+  newSet.Broadcast();
+  return true;
+}
+
 bool cDvbTuner::GetFrontendStatus(fe_status_t &Status, int TimeoutMs)
 {
   if (TimeoutMs) {
@@ -302,6 +317,10 @@ void cDvbTuner::Action(void)
         if (GetFrontendStatus(NewStatus, 10))
            Status = NewStatus;
         cMutexLock MutexLock(&mutex);
+        if (SendDiseqc) {
+           CHECK(ioctl(fd_frontend, FE_DISEQC_SEND_MASTER_CMD, &diseqc_cmd));
+           SendDiseqc=false;
+           }
         switch (tunerStatus) {
           case tsIdle:
                break;
@@ -669,6 +688,11 @@ bool cDvbDevice::HasLock(int TimeoutMs)
   return dvbTuner ? dvbTuner->Locked(TimeoutMs) : false;
 }
 
+bool cDvbDevice::SendDiseqcCmd(dvb_diseqc_master_cmd cmd)
+{
+  return dvbTuner->SendDiseqcCmd(cmd);
+}
+
 void cDvbDevice::SetTransferModeForDolbyDigital(int Mode)
 {
   setTransferModeForDolbyDigital = Mode;
diff -Naurp vdr-1.7.11/dvbdevice.h vdr-1.7.11-rotor/dvbdevice.h
--- vdr-1.7.11/dvbdevice.h	2010-01-04 14:07:12.000000000 +0000
+++ vdr-1.7.11-rotor/dvbdevice.h	2010-01-07 23:49:39.000000000 +0000
@@ -80,6 +80,7 @@ protected:
   virtual bool SetChannelDevice(const cChannel *Channel, bool LiveView);
 public:
   virtual bool HasLock(int TimeoutMs = 0);
+  virtual bool SendDiseqcCmd(dvb_diseqc_master_cmd cmd);
 
 // PID handle facilities
 
